src/react-components/rtc-debug-panel.js

```diff
@@ -0,0 +1,1011 @@
+  }
+  componentWillUnmount() {
+    this.props.store.removeEventListener("statechanged", this.storeUpdated);
+    document.body.removeEventListener("locale-updated", this.storeUpdated);
+  }
+
+  storeUpdated = () => {
+    this.forceUpdate();
+  };
+
+  render() {
+    return (
+      <CollapsablePanel title={"Device"} border={true} url={`${MEDIASOUP_DOC_BASE_URL}#Device`}>
+        <Prop propKey={`loaded`} propValue={`${this.props.data.loaded}`} />
+        {/* <Prop propKey={`codecs`} propValue={`${this.props.data.codecs}`} /> */}
+      </CollapsablePanel>
+    );
+  }
+}
+
+export class LogPanel extends Component {
+  static propTypes = {
+    log: PropTypes.array
+  };
+  componentDidMount() {
+    document.body.addEventListener("locale-updated", this.storeUpdated);
+  }
+  componentWillUnmount() {
+    document.body.removeEventListener("locale-updated", this.storeUpdated);
+  }
+
+  storeUpdated = () => {
+    this.forceUpdate();
+  };
+
+  colorForLevel(level) {
+    switch (level) {
+      case "error":
+        return "red";
+      case "warn":
+        return "orange";
+      default:
+        return "white";
+    }
+  }
+
+  render() {
+    const logLines = [];
+    this.props.log
+      ?.slice()
+      .reverse()
+      .forEach((log, index) => {
+        logLines.push(<span key={index << 2} className={classNames(styles.rtcLogTag)}>{`[${log.tag}] `}</span>);
+        logLines.push(
+          <span
+            key={(index << 2) | 0x01}
+            className={classNames(styles.rtcLogMsg)}
+            style={{ color: this.colorForLevel(log.level) }}
+          >
+            {log.msg}
+          </span>
+        );
+        logLines.push(<br key={(index << 2) | 0x02} />);
+      });
+    return (
+      <div className={classNames(styles.rtcLogContainer)}>
+        <CollapsablePanel title={"Log"}>
+          <p style={{ userSelect: "all" }}>{logLines}</p>
+        </CollapsablePanel>
+      </div>
+    );
+  }
+}
+
+export class CollapsablePanel extends Component {
+  static propTypes = {
+    children: PropTypes.node,
+    title: PropTypes.string,
+    border: PropTypes.bool,
+    row: PropTypes.bool,
+    wrap: PropTypes.bool,
+    url: PropTypes.string
+  };
+
+  collapse(evt) {
+    evt.target.classList.toggle("collapsed");
+    evt.target.parentElement?.nextSibling?.classList.toggle("collapsed");
+  }
+
+  openLink() {
+    const win = window.open(this.props.url, "_blank");
+    win.focus();
+  }
+
+  render() {
+    const rootClassName = this.props.border ? classNames(styles.rtpTile) : classNames(styles.rtcStatusPanel);
+    const flow = { flexFlow: (this.props.row ? "row " : "column ") + (this.props.wrap ? "wrap" : "nowrap") };
+    return (
+      <div className={rootClassName}>
+        <div style={{ display: "flex", flexFlow: "row nowrap", justifyContent: "center" }}>
+          {this.props.title && (
+            <button className={classNames(styles.collapsibleButton)} onClick={this.collapse.bind(this)}>
```

Comment:
You can avoid having to bind these in every render by either binding them in the constructor, or defining them using the public class field syntax: `collapse = () => {`

Reply:
Fixed, thanks for that.